File: cpp/spiral_matrix.cc
/*
 * Filename: spiral_matrix.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Mon 27 Jul 2015 11:26:56 AM EDT
 */

#include <iostream>
#include <vector>

using namespace std;

class Solution {
	public:
		vector<int> spiralOrder(vector<vector<int> >& matrix) {
			if (matrix.size() == 0 || matrix[0].size() == 0) {
				return res;
			}

			vector<int> res;

			int m = matrix.size();
			int n = matrix[0].size();
			int layer = min(m, n) / 2;

			for (vector<int>::size_type i = 0; i < layer; i++) {
				for (vector<int>::size_type j = i; j <= n-i-2; j++) {
					res.push_back(matrix[i][j]);
				}

				for (vector<int>::size_type j = i; j <= m-i-2; j++) {
					res.push_back(matrix[j][n-i-1]);
				}

				for (vector<int>::size_type j = i; j <= n-i-2; j++) {
					res.push_back(matrix[m-i-1][n-j-1]);
				}

				for (vector<int>::size_type j = i; j <= m-i-2; j++) {
					res.push_back(matrix[m-j-1][i]);
				}
			}

			if (m > n && (n & 0x1)) {
				for (vector<int>::size_type i = layer; i < m - layer; i++) {
					res.push_back(matrix[i][n/2]);
				}
			} else if (n >= m && (m & 0x1)) {
				for (vector<int>::size_type i = layer; i < n - layer; i++) {
					res.push_back(matrix[m/2][i]);
				}
			}

			return res;
		}
};

File: java/SpiralMatrix.java
/*
 * Filename: SpiralMatrix.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 15 May 2015 08:06:53 PM EDT
 */

public class Solution {
	public List<Integer> spiralOrder(int[][] matrix) {
		List<Integer> res = new ArrayList<Integer>();

		if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
			return res;
		}

		int m = matrix.length;
		int n = matrix[0].length;
		int eend = Math.min(m, n);
		int end = eend / 2;

		for (int i = 0; i < end; i++) {
			for (int j = i; j < n-i-1; j++) {
				res.add(matrix[i][j]);
			}

			for (int j = i; j < m-i-1; j++) {
				res.add(matrix[j][n-i-1]);
			}

			for (int j = n-i-1; j > i; j--) {
				res.add(matrix[m-i-1][j]);
			}

			for (int j = m-i-1; j > i; j--) {
				res.add(matrix[j][i]);
			}
		}

		if (eend % 2 == 1) {
			if (m >= n) {
				for (int i = end; i < m - end; i++) {
					res.add(matrix[i][end]);
				}
			} else {
				for (int i = end; i < n - end; i++) {
					res.add(matrix[end][i]);
				}
			}
		}

		return res;
	}
}
